live-build/misc/upgrade-scripts/upgrade-container

```diff
@@ -273,12 +273,79 @@ function run() {
 	systemd-run --machine="$CONTAINER" --quiet --wait --pipe -- "$@"
 }
 
+function get_bootloader_devices() {
+	#
+	# When installing/updating the bootloader during upgrade, we
+	# need to determine which devices are being used as bootloader
+	# devices. We determine this by listing the devices used by the
+	# rpool. Additionally, we have to filter out devices that could
+	# be attached to the rpool, but would never be used for the
+	# bootloader. Finally, we need to strip off any parition
+	# information, since we want to install the bootloader directly
+	# to the device, rather than to a partition of the device.
+	#
+	zpool list -vH rpool |
+		awk '! /rpool|mirror|replacing|spare/ {print $1}' |
+		sed 's/[0-9]*$//'
+}
+
+function convert_to_bootfs_cleanup() {
+	for dir in /dev /proc /run /sys; do
+		umount "/var/lib/machines/${CONTAINER}${dir}" ||
+			warn "failed to umount '$dir' from '$CONTAINER'"
+	done
+}
+
+function convert_to_bootfs() {
+	trap convert_to_bootfs_cleanup EXIT
+
+	for dir in /dev /proc /run /sys; do
+		mount --bind "$dir" "/var/lib/machines/${CONTAINER}${dir}" ||
```

Comment:
mount `--bind` bounds non-recursively, which can mess up some devices. I think we should use `--rbind` and then `--make-rslave` to make sure we do not unmount those devices from the root environement.

Also, do we actually need to mount all of those devices?

Reply:
> I think we should use `--rbind` and then `--make-rslave` to make sure we do not unmount those devices from the root environment.

Can you give me the commands that I should use to mount, and then the commands I should use to unmount?

I'm willing to try it out, I'm just not sure how to do it cleanly without the ugly complexity that we do for the raw image creation (which I'd like to avoid if possible), [here](https://github.com/delphix/appliance-build/blob/master/live-build/misc/live-build-hooks/90-raw-disk-image.binary#L147) and [here](https://github.com/delphix/appliance-build/blob/master/live-build/misc/live-build-hooks/90-raw-disk-image.binary#L133).

Reply:
Hmm, so I think the way it's done in the links you provided might cause some issues as we didn't specify `--make-rslave`, and so unmounting the child fs in the bind-mount would also unmount the child fs in the original mount. 

Here's an example:
```
	for path in /proc /sys /dev /run; do
		mount --rbind ${path} ${NROOT}${path}
		mount --make-rslave ${NROOT}${path}
	done
```
```
	for path in /proc /sys /dev /run; do
		umount -R ${NROOT}${path}
	done
```
Note that even if we use bind instead of rbind, we might still need to use `--make-slave`.

Reply:
Interesting.. no need to do this complicated `mount | grep | tac | awk | xargs` nonsense? That seems simple enough.. I'll give it a try.